client/core/types.go

```diff
@@ -735,6 +736,14 @@ func (a *dexAccount) ID() account.AccountID {
 	return a.id
 }
 
+// isRegistered is true if this account has a feeCoin. Every registered account
+// must have a feeCoin set.
+func (a *dexAccount) isRegistered() bool {
+	a.keyMtx.RLock()
+	defer a.keyMtx.RUnlock()
+	return len(a.feeCoin) > 0
```

Comment:
`a.feeCoin` is not an good indicator of an unconfigured `dexAccount` (bonds, but also payment status is not a direct indicator of the existence of an account identity/key).

What are the characteristics of a `dexAccount` created by the `AddDEX` method?  I think the answer is that it has not been "initialized" via `setupCryptoV2`, i.e. the fields guarded by `keyMtx` are still zero-valued.  The fee, bond, and auth related fields guarded by `authMtx` are not what we are interested in I think.

The previous `(*Core).isRegistered` had the appropriate check, although now it's expected to have a `dexConnection` in the map with a `dexAccount` with no configured identity.

Reply:
That was my first approach too, checking `a.id`, or `a.encKey`/`a.privKey`. But realized that those fields would be generated (by `DiscoverAccount -> setupCryptoV2`) if a registration is attempted but not completed. Thought about zeroing those fields if `DiscoverAccount` does not find a paid account and `Register` does not pay the required fee, but I worry things may fall through the crack with that approach. At least for a brief period, those keys won't be empty and the account will be thought to be registered (by `a.isRegistered`).

Another thought was to NOT use the dexConnection added by `AddDEX` in `DiscoverAccount` and `Register`, to ensure that that dexAccount has empty keys but doing this will cause `DiscoverAccount` and `Register` to create a new dexConnection which would override the initial one in the conns map if all goes well. Going this route, I'd have to properly close the existing connection (stop book subscriptions etc) before replacing in the conns map. Dislike this approach primarily because we'd be setting up (and potentially promoting) a new connection when we have a connection that is almost g2g.

So to answer directly,
> What are the characteristics of a `dexAccount` created by the `AddDEX` method?

It has not been initialized via setupCryptoV2, but could be at a later time even though the dexAccount is still not ready for trading.

Perhaps, I should rename `a.isRegistered` to `a.readyToTrade` or `a.canTrade`? But is it really registered if the keys are generated by setupCryptoV2 but the fee isn't paid yet? I agree this check will need to change with bonds, but pre-bonds, it does make sense to me to rely more on whether a fee has been paid or not rather than on whether keys have been generated.

> The fee, bond, and auth related fields guarded by `authMtx` are not what we are interested in, I think (at least in most of the places you are using `a.isRegistered`).

I'll take another look at the call sites to confirm, but could any of such callers only care about the presence of generated keys even if the account may not have paid the required fee?

Reply:
Things have changed quite a bit recently, and yes, it really is the account keys/identity that matters.  `feeCoin` is a deprecated field, and it is expected to be empty if you create an account with a bond, and the field will eventually be removed.  But _even if_ we only had the legacy registration fee system, it's still not the appropriate indicator of whether a `dexConnection` and it's associated `dexAccount` is a **view-only connection** (let's call the opposite an "active" conn/account).

You can have a conn/account that is converted to "active" from "view-only" that is not fully registered.  I will try to give reasoning for this.

> That was my first approach too, checking `a.id`, or `a.encKey`/`a.privKey`. But realized that those fields would be generated (by `DiscoverAccount -> setupCryptoV2`) if a registration is attempted but not completed. Thought about zeroing those fields if `DiscoverAccount` does not find a paid account and `Register` does not pay the required fee, but I worry things may fall through the crack with that approach. At least for a brief period, those keys won't be empty and the account will be thought to be registered (by `a.isRegistered`).

Calling either `Register` or `DiscoverAccount` should be understood to permanently convert a "view-only" account/connection to an "active"/"trading" account/connection (wording TBD).  This should be documented, but it is a reasonable expectation for the caller of these methods.  There's not much ambiguity in the intent of the caller of these methods.

If `Register` is "not completed" for whatever reason, that does not change the fact that the dex/account has been setup and is destined for activation on the DEX via either the legacy fee or posting bonds.  If we want it to revert to a "view-only" account on error, then yes, we'd have to zero this stuff out.  But depending on the failure, there is or should be handling like `reFee` to complete the registration without calling again, but I don't think that's central to the issue.

> > What are the characteristics of a `dexAccount` created by the `AddDEX` method?
> 
> It has not been initialized via setupCryptoV2, but could be at a later time even though the dexAccount is still not ready for trading.
> 
> Perhaps, I should rename `a.isRegistered` to `a.readyToTrade` or `a.canTrade`? But is it really registered if the keys are generated by setupCryptoV2 but the fee isn't paid yet? I agree this check will need to change with bonds, but pre-bonds, it does make sense to me to rely more on whether a fee has been paid or not rather than on whether keys have been generated.

Bond are *now*.  It's not a future concept anymore.  The browser UI doesn't do it yet, but you can create a new account with a bond via `dexcctl --simnet -p a postbond 127.0.0.1:17273 1000000000 42` today and `feeCoin` will always be empty.

Depending on the context, it does not matter how you registered (fee vs. bond) or the current account status (tier), just that you've configured the account and it's now intended for trading (`!viewOnly`) regardless of whether your account is tier>0 and thus "readyToTrade".

So, regarding "`a.readyToTrade` or `a.canTrade`", these aren't always the pertinent concepts.  Sometimes we want to know if the `dexConnection`+`dexAccount` a "view-only" or "trading" connection, and sometimes we want to know if tier>0 and thus can place orders, and sometimes like in `Register` we want to know if there is already a `feeCoin` (but the legacy fee is now a deprecated account creation method).  `(*Core).PostBond` does not care at all if `feeCoin` is set; only if the account identity and keys are initialized (not a view-only conn).  `authDEX` doesn't care if there's a `feeCoin` or live bonds, but it will fail if there's no `feeCoin` and the account's very first bond hasn't yet been activated (when server first makes the account entry in the DB), but after that `authDEX` will succeed with no `feeCoin` and even if there are no more active bonds -- tier will just be <1.

> > The fee, bond, and auth related fields guarded by `authMtx` are not what we are interested in, I think (at least in most of the places you are using `a.isRegistered`).
> 
> I'll take another look at the call sites to confirm, but could any of such callers only care about the presence of generated keys even if the account may not have paid the required fee?

We have to forget about the registration fee.  But even if that were the only method of account registration for the forseable future, yes, we would still not be concerned with the `feeCoin` but rather if the `dexAccount` were given an identity.  Only `Register` cares about `feeCoin`.  `(*dexConnection).exchangeInfo` should show the account ID if it's non-zero.  `AddDEX` should fail if there's a `dexAccount` with just the keys set because that could be a bonded account with no fee coin.  `DiscoverAccount` should fail if there's a `dexAccount` with keys *and* `authed()`, otherwise it's OK to try the `authDEX` loop.  `initializeDEXConnections` and `handleReconnect` can attempt `authDEX` for any account with keys for the same reasons.  `Trade` (via `connectedDEX`), etc. should only fail if tier<1.